atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -417,7 +426,12 @@ public InetSocketAddress apply(String endpoint) {
             }
 
             try {
-                return hostPool.runWithPooledResource(f);
+                V result = hostPool.runWithPooledResource(f);
+                InetSocketAddress host = hostPool.getHost();
+                if (blacklistedHosts.containsKey(host)) {
```

Comment:
So you're saying that since this faulty server could serve simple requests but not more complex/slower ones, that this would potentially cause it to continuously add and remove to/from the blacklist? I don't think concurrency would affect this because the health check is happening in a single thread regardless. However, I do see the thrashing on a faulty-but-still-sort-of-alive server argument if that's what you mean.